core/packages/contracts/contracts/SovereignTreasury.sol

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+import "./Vault.sol";
+
+// TODO: transfer ownership from deployer to inbound channel
+contract SovereignTreasury is Ownable {
+    Vault public vault;
+
+    constructor(Vault _vault) {
+        vault = _vault;
+    }
+
+    // Handle a message from the bridge.
+    function handle(bytes32 _sovereignID, bytes calldata _message) external onlyOwner {
+        Message memory message = abi.decode(_message, (Message));
+
+        if (message.action == Action.Withdraw) {
+            WithdrawPayload memory payload = abi.decode(message.payload, (WithdrawPayload));
+
+            transfer(_sovereignID, payload.recipient, payload.amount);
+        }
+    }
+
+    // Deposit ETH into a sovereign account. Permissionless.
+    function deposit(bytes32 _sovereignID) external payable {
+        vault.deposit{ value: msg.value }(_sovereignID);
+    }
+
+    function transfer(bytes32 sovereignID, address payable recipient, uint256 amount) private {
```

Comment:
iirc, passing around reference types in `memory` like this is more expensive than passing value types, but the abstraction makes more sense & we can optimise later.

Reply:
I believe it's okay when passing them to private/internal functions which are optimized to jumps and data is not copied. It might just be the same when optimized.